object-model/copy-semantics.md

```diff
@@ -0,0 +1,153 @@
+# C++ object model: copy semantics
+_Skeleton instructions are typeset in italic text._
+
+## Overview
+
+_Provides a short natural language abstract of the module’s contents._
+_Specifies the different levels of teaching._
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>understanding how and when are copies made</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>implementing user-defined copy operations</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td>special cases: copy elision</td>
+  </tr>
+</table>
+
+## Motivation
+
+_Why is this important?_
+_Why do we want to learn/teach this topic?_
+
+Copy semantics allows the user to determine how objects of a class get replicated and interact on a value level.
+
+## Topic introduction
+
+_Very brief introduction to the topic._
+
+Explains when and how objects are copied.
+
+## Foundational knowledge: How and when are copies made
+
+### Background/Required Knowledge
+
+A student is able to:
+* explain what a C++ type is?               [[C++ object model: types]][1]
+* explain what an object is?                [[C++ object model: objects]][2], [[C++ object model: constant objects]][3]
+* define and understand class invariants?
+
+It helps when a student is able to:
+* use move semantics                        [[C++ object model: move semantics]][4]
+* explain special member functions          [[C++ object model: special member functions]][5]
+
+### Student outcomes
+
+_A list of things "a student should be able to" after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. explain what copy semantics accomplish
+  * establishing "equivalent" object state in another object
+2. explain difference between copying a reference and copying a value\*
+3. explain where copies are made
```

Comment:
As Amir brought up on the call, I think it's unclear whether this unit is supposed to permit teaching students about copy semantics _without mentioning how to customize them_, or not.

It is possible to say a fair bit of stuff about C++'s relatively novel "copy semantics" before bringing up customizable copy operations at all. E.g., when you define `Widget y = x;`, in Java or Python or JavaScript you'd implicitly get a _pointer_ pointing at the same object; but in C++ you get a _whole new Widget_, just like `int y = x;` in any of these languages. So "copies are made" whenever you pass a value to a function, or return a value, or initialize one variable with a copy of another, or throw, or catch (by value), or in a `for (auto elt : range)` loop, or probably other places I'm forgetting. C++ looooves copies. You can explain all this (and the behavior of Rule-of-Zero defaulted copy operations) without even mentioning special members.

The advanced version would then be to explain that it is possible to overload the constructor with a signature that says "given (a reference to another Widget), make me a new Widget," i.e., a copy constructor; and to explain why you might want to do that. This requires knowing about constructors, knowing about references, etc.; and would be taught right alongside destructors and how-to-overload-the-assignment-operator.

Reply:
Should we specifically add wording that says we only what teachers in this stage to explain: Where copies are made without referring to user defined copy operations?
The student outcomes section seems to be the wrong place. Which section would be the best place to clarify this? @jcvw Do you have an idea?